### ISSUE TITLE
What Python installation/PaaS?

### ISSUE THREAD

**Issue by SJaffa** [UoMResearchIT/programming_with_python#26]
Last time we used a VM with Python already installed but this might not be an option in the future?

Decisions to make:
- Do we (a) get everyone to install it on their own machines or (b) provide a platform with it preinstalled?
- If (a), what installation method, works for all OSs, need extra pre-workshop setup session, extra helpers to start
- If (b), what service, who/how hosting, who pays?

**Comment by SJaffa** [comment]
@douglowe what solution did we use on the course in the end and was it suitable? Is that what we want to go with long term?

**Comment by douglowe** [comment]
We used the VM that we have used for the last 3-4 years.

We should investigate the jupyter lab servers more over the next 2 months, and see if they are suitable though. It might be nice for us to use a central solution instead.

**Comment by SJaffa** [comment]
Current solution: ITS-owned Jupyter Hub located at https://jupyter.its.manchester.ac.uk/

We requested an increase in memory allocation per user which should allow people to run the conda installations, it might be slow if they all do it at once but we will see on the day.

We need to give themget them to download the data and upload it to the jupyter hub.

**Comment by SJaffa** [comment]
In the terminal in Jupyter hub:

listing and creating conda environments works, but before acivating you need to do `conda init bash`. This advises you to close the terminal and open it which loses the bashrc, so instead you can keep the terminal open and run `source ~/.bashrc` to initialise conda, then activate it

- [ ] How do we get the notebook to use the environment?

**Comment by SJaffa** [comment]
Could run scripts form the terminal with `conda run -n py38 python my_script.py` [(source)](https://stackoverflow.com/questions/72145492/conda-init-without-closing-the-current-shell)

**Comment by SJaffa** [comment]
All the libraries we use in the course are preinstalle din the Jupyter Hub (I did Ctrl-f "import" and checked they all run), so we don't _need_ to get the environment working for the notebook, it jsut seems a bit disjointed to teach them and then never use it...

**Comment by douglowe** [comment]
I agree it is now disjointed - the central team were a bit too helpful in installing all the libraries we needed for this course.

Ideally we'll split out the packaging episode into a dedicated packaging lesson at some point. For now perhaps you could simply move this lesson to the end of the day, so that it doesn't disrupt the flow of the rest of the episodes.

**Comment by andrewgait** [comment]
I agree with @douglowe that the temporary solution here should be to move the `conda` chapter to the end of the day.

**Comment by andrewgait** [comment]
We used the central Jupyter instance with no issues yesterday (as far as I could see, the biggest issue was it took a while to load up initially for some people).  I would probably argue that this issue can be closed, and if we need to reconsider what we use in the future then a new issue for that should be made.  @Scottan any thoughts?

**Comment by Scottan** [comment]
Yeah, I think this can be closed. The Jupyter server worked well, the main problem was that it seemed to struggle with lots of people logging on at the same time at the start of the course, but that was largely solved by waiting a bit then refreshing.

If we need to revisit this we can open a new issue. There might need be an action in the main training repo to ask for more resource for the server from infra to see if that helps with its performance? (it might just need a clean-out as it has been used for a number of course already)
